python/private/pypi/dependency_specifier_flag.bzl

```diff
@@ -0,0 +1,84 @@
+load("//python/private:toolchain_types.bzl", "TARGET_TOOLCHAIN_TYPE")
+load(":pep508_evaluate.bzl", "evaluate")
+
+def _impl(ctx):
+    env = {}
+
+    runtime = ctx.toolchains[TARGET_TOOLCHAIN_TYPE].py3_runtime
+    if runtime.interpreter_version_info:
+        version_info = runtime.interpreter_version_info
+        env["python_version"] = "{v.major}.{v.minor}".format(v = version_info)
+        full_version = format_full_version(version_info)
+        env["python_full_version"] = full_version
+        env["implementation_version"] = full_version
+    else:
+        env["python_version"] = get_flag(ctx.attr._python_version)
+        full_version = get_flag(ctx.attr._python_full_version)
+        env["python_full_version"] = full_version
+        env["implementation_version"] = full_version
+
+    # We assume cpython if the toolchain doesn't specify because it's most
+    # likely to be true.
+    env["implementation_name"] = runtime.implementation_name or "cpython"
+
+    env["os_name"] = struct()
+    env["platform_machine"] = struct()
+    env["platform_python_implementation"] = struct()
```

Comment:
This could be `CPython` if `cpython` is implementation name. We could probably have that as a flag/or attach it to the `runtime`.

Reply:
I dug through peps a bit. From what i can tell, sys.implementation.name was created to replace platform.python_implementation. It turns out the `platform` uses regex parsing of the `sys.version` string to determine its values :astonished: .

The docs say there are 4 possible values, 2 of which are essentially defunct (IronPython, Jython). The other two are CPython and PyPy.

I just added an if condition to transform cpython to CPython. If we know the sys.implementation.name value that pypy uses, we could special case that, too.